modules/dists/SparseBlockDist.chpl

```diff
@@ -117,6 +116,64 @@ class SparseBlockDom: BaseSparseDom {
     dsiAdd((ind,));
   }
 
+  proc dsiBulkAdd(inds: [] index(rank, idxType), isSorted=false, isUnique=false,
+      preserveInds=true){
+  
+    if !isSorted && preserveInds {
+      var _inds = inds;
+      bulkAdd_help(_inds, isSorted, isUnique); 
+    }
+    else {
+      bulkAdd_help(inds, isSorted, isUnique);
+    }
+  }
+
+  // Tried to put this record in the function and the if statement, but got a
+  // segfault from the compiler.
+  record TargetLocaleComparator {
+    proc key(a: index(rank, idxType)) { 
+      return (dist.targetLocsIdx(a), a); 
+    }
+  }
+
+  proc bulkAdd_help(inds: [] index(rank,idxType), isSorted=false, isUnique=false) {
+
+    if !isSorted {
+
+      // without _new_ record functions throw null deref. It doesn't seem to be 
+      // able to deref _outer_ ? I think it has something to do with memory
+      // allocation for classes vs records, but I cannot explain
+      var comp = new TargetLocaleComparator;
+      sort(inds, comparator=comp);
+    }
+
+    var active : atomic int;
+
+    var firstIndex = inds.domain.low;
+    var curLoc = dist.targetLocsIdx(inds[inds.domain.low]);
+
+    for i in inds.domain {
+      const _tmpLoc = dist.targetLocsIdx(inds[i]);
+      if _tmpLoc != curLoc {
+        spawnBulkAdd(firstIndex..i-1, curLoc);
+        curLoc = _tmpLoc;
+        firstIndex = i;
+      }
+    }
+    spawnBulkAdd(firstIndex..inds.domain.high, curLoc);
+
+    active.waitFor(0);
```

Comment:
I'm also very curious about the performance implications of doing this serially vs using the begins. Is that something you measured? If not, I think it would be good to measure in some way.